packages/e2e/lib/e2e.dart

```diff
@@ -128,9 +137,57 @@ class E2EWidgetsFlutterBinding extends LiveTestWidgetsFlutterBinding {
   /// The default value is `null`.
   Map<String, dynamic> reportData;
 
+  /// Tests will put the action requests from WebDriver to this pipe.
+  Completer<WebDriverAction> webDriverActionPipe = Completer<WebDriverAction>();
+
+  /// Updated when WebDriver completes the request by the test method.
+  ///
+  /// For example, a test method will ask for a screenshot by calling
+  /// `takeScreenshot`. When this screenshot is taken [driverActionComplete]
+  /// will complete.
+  Completer<bool> driverActionComplete = Completer<bool>();
+
+  final Mutex webDriverCommandLock = new Mutex();
+
+  /// Takes screenshot using WebDriver screenshot command.
+  ///
+  /// Only works on Web when tests are run via `flutter driver` command.
+  ///
+  /// See: https://www.w3.org/TR/webdriver/#screen-capture
+  Future<void> takeScreenshot(String screenshot_name) async {
+    if (!kIsWeb) {
+      throw StateError('Web driver commands are only avaliable on web');
+    }
+    await _webDriverCommand(WebDriverAction.screenshot(screenshot_name));
+  }
+
+  Future<void> _webDriverCommand(WebDriverAction command) async {
+    await webDriverCommandLock.acquire();
+
+    try {
+      webDriverActionPipe.complete(Future.value(command));
+      try {
+        final bool awaitCommand = await driverActionComplete.future;
+        if (!awaitCommand) {
+          throw Exception('Web Driver Command failed: ${command.type}');
+        }
+      } catch (e) {
+        throw Exception(
+            'Web Driver Command failed: ${command.type} with ' 'exception $e');
+      }
+    } finally {
+      // Reset the completer and release the lock.
+      driverActionComplete = Completer<bool>();
```

Comment:
Why are we reusing this? Why can't we just make a new one in this function and avoid the whole mutex business?

Reply:
The mutex makes sure only one methods executes this method:
```
    await webDriverCommandLock.acquire();

    try {
      webDriverActionPipe.complete(Future.value(command));
      try {
        final bool awaitCommand = await driverActionComplete.future;
        if (!awaitCommand) {
          throw Exception('Web Driver Command failed: ${command.type}');
        }
      } catch (e) {
        throw Exception(
            'Web Driver Command failed: ${command.type} with ' 'exception $e');
      }
    } finally {
      // Reset the completer and release the lock.
      driverActionComplete = Completer<bool>();
      webDriverCommandLock.release();
    }
```

I'm trying to avoid a race condition on `webDriverActionPipe.complete(Future.value(command));` for example.

Reply:
I removed the mutex: https://github.com/flutter/plugins/pull/2904#discussion_r475828717